# Natural Language Processing Project

# Multi-Label Categorical Classification of English Quotes

<b>Rish (Rishabh) Kaushick</b> <br>
College of Engineering <br>
Northeastern University <br>
Toronto, ON <br>
kaushick.r@northeastern.edu

# Downloading The Dataset From Hugging Face

In [1]:
from huggingface_hub import hf_hub_download
import pandas as pd

REPO_ID = "Abirate/english_quotes"
FILENAME = "quotes.jsonl"

# since the data is in jsonl format, we read it using read_json() function with lines=True
dataset = pd.read_json(hf_hub_download(repo_id=REPO_ID, filename=FILENAME, repo_type="dataset"), lines=True)


In [2]:
dataset

quote  \
0        “Be yourself; everyone else is already taken.”   
1     “I'm selfish, impatient and a little insecure....   
2     “Two things are infinite: the universe and hum...   
3                      “So many books, so little time.”   
4     “A room without books is like a body without a...   
...                                                 ...   
2503  “Morality is simply the attitude we adopt towa...   
2504  “Don't aim at success. The more you aim at it ...   
2505  “In life, finding a voice is speaking and livi...   
2506  “Winter is the time for comfort, for good food...   
2507                      “Silence is so freaking loud”   

                     author                                               tags  
0               Oscar Wilde  [be-yourself, gilbert-perreira, honesty, inspi...  
1            Marilyn Monroe  [best, life, love, mistakes, out-of-control, t...  
2           Albert Einstein  [human-nature, humor, infinity, philosophy, sc...  
3               Frank Zappa                                     [books, humor]  
4     Marcus Tullius Cicero                              [books, simile, soul]  
...                     ...                                                ...  
2503           Oscar Wilde,                             [morality, philosophy]  
2504      Viktor E. Frankl,                               [happiness, success]  
2505           John Grisham                               [inspirational-life]  
2506          Edith Sitwell                            [comfort, home, winter]  
2507          Sarah Dessen,    [just-listen, loud, owen, sara-dessen, silence]  

[2508 rows x 3 columns]

# Exploratory Data Analysis

In [3]:
dataset.head(10)

quote                 author  \
0     “Be yourself; everyone else is already taken.”            Oscar Wilde   
1  “I'm selfish, impatient and a little insecure....         Marilyn Monroe   
2  “Two things are infinite: the universe and hum...        Albert Einstein   
3                   “So many books, so little time.”            Frank Zappa   
4  “A room without books is like a body without a...  Marcus Tullius Cicero   
5  “Be who you are and say what you feel, because...      Bernard M. Baruch   
6  “You've gotta dance like there's nobody watchi...      William W. Purkey   
7  “You know you're in love when you can't fall a...              Dr. Seuss   
8  “You only live once, but if you do it right, o...               Mae West   
9  “Be the change that you wish to see in the wor...         Mahatma Gandhi   

                                                tags  
0  [be-yourself, gilbert-perreira, honesty, inspi...  
1  [best, life, love, mistakes, out-of-control, t...  
2  [human-nature, humor, infinity, philosophy, sc...  
3                                     [books, humor]  
4                              [books, simile, soul]  
5  [ataraxy, be-yourself, confidence, fitting-in,...  
6  [dance, heaven, hurt, inspirational, life, lov...  
7                     [dreams, love, reality, sleep]  
8                                      [humor, life]  
9  [action, change, inspirational, philosophy, wish]

In [4]:
# setting random_state in the random sample, so that each time the same value is presented
dataset.sample(5, random_state=23)

quote             author  \
2062  “Why are you worrying about YOU-KNOW-WHO, when...      J.K. Rowling,   
1078  “I am not sure that I exist, actually. I am al...  Jorge Luis Borges   
1176  “If you're lonely when you're alone, you're in...   Jean-Paul Sartre   
414   “You educate a man; you educate a man. You edu...      Brigham Young   
508   “There is neither happiness nor misery in the ...    Alexandre Dumas   

                                   tags  
2062                            [humor]  
1078                             [life]  
1176             [loneliness, solitude]  
414   [education, feminism, men, women]  
508       [hope, inspirational, wisdom]

In [7]:
# Looking for how many unique values are present in the author column
len(dataset['author'].unique())

880

In [8]:
dataset['tags'].value_counts()

tags
[love]                                                                                                                                                                                                                          75
[inspirational]                                                                                                                                                                                                                 63
[humor]                                                                                                                                                                                                                         37
[life]                                                                                                                                                                                                                          34
[writing]                                                                              

The above does not help us much because one row could have multiple categories in the list. <br>
Therefore, in order to see how many unique tags are there, we might need to iterate through the entire pandas dataframe, and store them in a Python set.

In [11]:
# iterating through the values in the pandas dataframe
tags = dataset['tags']
print(tags)
print(type(tags))

0       [be-yourself, gilbert-perreira, honesty, inspi...
1       [best, life, love, mistakes, out-of-control, t...
2       [human-nature, humor, infinity, philosophy, sc...
3                                          [books, humor]
4                                   [books, simile, soul]
                              ...                        
2503                               [morality, philosophy]
2504                                 [happiness, success]
2505                                 [inspirational-life]
2506                              [comfort, home, winter]
2507      [just-listen, loud, owen, sara-dessen, silence]
Name: tags, Length: 2508, dtype: object
<class 'pandas.core.series.Series'>


In [13]:
# Iterting though the pandas Series object and adding each value in a Set
# The benefit of the set is that only the unique values will be added, therefore all duplicates will be removed
unique_tags = set()

# row_tags is the list of tags in each row
# we can see that sometimes a single row has multiple tags, therefore, we will nest another for loop inside which reads individual tags
for row_tags in tags:
    for tag in row_tags:
        unique_tags.add(tag)

print(unique_tags)

{'style', 'gullibility', 'bad-day', 'mindful', 'inspirational', 'together', 'choices', 'opinions', 'trap', 'brokenness', 'quotes', 'giving-in', 'stereotypes', 'swear', 'fudge', 'moi', 'crescendo', 'lookingforalaska', 'wormtail', 'rulers', 'last-words', 'greatness', 'independent-thought', 'science', 'diary-entry', 'positivity', 'separation', 'thirst', 'hindsight', 'yesterday', 'talent', 'worry', 'challenges', 'night', 'personal-responsibility', 'fair', 'nicholas-sparks', 'beatrice', 'Ø\xadب', 'clichés', 'gandhi', 'thankful', 'jordan-kyle', 'deceit', 'letting-go-of-the-past', 'proactivity', 'harry-potter', 'school', 'broken-families', 'self-containment', 'longings', 'amor-deliria-nervosa', 'john-green', 'enterprise', 'censorship', 'cleanliness', 'right-way', 'revolution', 'life-and-living', 'misogyny', 'prejudice', 'dreamer', 'haunting', 'thoughts', 'think-different', 'water', 'ocean', 'changes', 'demons', 'classic-insult', 'smile', 'inspirational-quote', 'integrity', 'stupidity', 'hon

In [14]:
len(unique_tags)

2161

There are a total of 2,161 tags. This is too many if we want this to be our target variable.

In [20]:
# Let me create a Python dictionary to store the tag and it's corresponding frequency throughout this dataset.
tag_freq = {}

for row_tag in tags:
    for tag in row_tag:
        if tag in tag_freq:
            tag_freq[tag] += 1
        else:
            tag_freq[tag] = 1

print(tag_freq)

{'be-yourself': 16, 'gilbert-perreira': 1, 'honesty': 12, 'inspirational': 319, 'misattributed-oscar-wilde': 2, 'quote-investigator': 1, 'best': 1, 'life': 295, 'love': 327, 'mistakes': 10, 'out-of-control': 1, 'truth': 61, 'worst': 1, 'human-nature': 9, 'humor': 254, 'infinity': 4, 'philosophy': 60, 'science': 26, 'stupidity': 7, 'universe': 6, 'books': 141, 'simile': 8, 'soul': 12, 'ataraxy': 4, 'confidence': 11, 'fitting-in': 1, 'individuality': 21, 'misattributed-dr-seuss': 1, 'those-who-matter': 1, 'dance': 5, 'heaven': 2, 'hurt': 8, 'sing': 2, 'dreams': 31, 'reality': 17, 'sleep': 7, 'action': 7, 'change': 34, 'wish': 2, 'character': 11, 'sirius-black': 2, 'friends': 28, 'friendship': 63, 'misattributed-albert-camus': 1, 'lies': 12, 'lying': 6, 'memory': 9, 'friend': 3, 'knowledge': 17, 'enemies': 4, 'forgiveness': 10, 'strategy': 3, 'carpe-diem': 7, 'education': 22, 'learning': 15, 'darkness': 7, 'drive-out': 1, 'hate': 10, 'light': 11, 'peace': 14, 'music': 31, 'intelligence': 

In [22]:
# Let us sort the dictionary in descending order of the frequencies:
tag_freq_sorted = sorted(tag_freq.items(), key=lambda x:x[1], reverse=True)
print(tag_freq_sorted)

[('love', 327), ('inspirational', 319), ('life', 295), ('humor', 254), ('books', 141), ('inspiration', 87), ('reading', 86), ('happiness', 68), ('writing', 67), ('inspirational-quotes', 65), ('friendship', 63), ('wisdom', 62), ('truth', 61), ('philosophy', 60), ('optimism', 48), ('living', 48), ('life-quotes', 48), ('death', 47), ('hope', 46), ('poetry', 45), ('romance', 45), ('positive-affirmation', 41), ('positive-thinking', 41), ('women', 40), ('life-lessons', 40), ('optimistic', 40), ('positive', 40), ('positive-life', 40), ('inspire', 38), ('inspiring', 38), ('courage', 37), ('life-and-living', 36), ('change', 34), ('pain', 34), ('funny', 34), ('inspirational-life', 34), ('relationships', 33), ('religion', 32), ('dreams', 31), ('music', 31), ('fear', 31), ('jace-wayland', 29), ('friends', 28), ('strength', 28), ('freedom', 28), ('god', 28), ('inspirational-attitude', 28), ('motivational', 27), ('loneliness', 27), ('science', 26), ('beauty', 26), ('time', 25), ('motivation', 25), (

# Naive Bayes (Baseline Model)

# Recurrant Neural Network (RNN) Model

# Transformer Based Model?